src/generate/rust.rs

```diff
@@ -1492,3 +813,540 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
+                put!(out, "},");
+            }
+        }
+        put!(out, "}");
+    } else {
+        put!(out, "#[allow(non_camel_case_types)]");
+        put!(out, "pub struct ", name, "<'a, 'i: 'a, I: 'a + ::gll::runtime::Input> {");
+        for (field_name, paths) in &rule.fields {
+            let refutable = rule.rule.field_pathset_is_refutable(paths);
+            put!(out, "pub ", field_name, ": ");
+            if refutable {
+                put!(out, "Option<");
+            }
+            put!(out, "Handle<'a, 'i, I, ", rule.rule.field_pathset_type(paths), ">");
+            if refutable {
+                put!(out, ">");
+            }
+            put!(out, ",");
+        }
+        if rule.fields.is_empty() {
+            put!(out, "_marker: PhantomData<(&'a (), &'i (), I)>,");
+        }
+        put!(out, "}");
+    }
+
+    // Implement `Debug`;
+    rule_debug_impls(name, &rule, variants.as_ref().map(|x| &**x), out);
+
+    put!(out, "impl<'a, 'i, I: ::gll::runtime::Input> ", name, "<'a, 'i, I> {"); // -- start impl
+    if let Some(variants) = &variants {
+        for Variant { rule, name: variant_name, fields } in variants {
+            put!(out, "
+                #[allow(non_snake_case)]
+                fn ", variant_name, "_from_sppf(
+                    parser: &'a ::gll::runtime::Parser<'i, _P, _C, I>,
+                    _node: ParseNode<'i, _P>,
+                    _r: traverse!(typeof(ParseNode<'i, _P>) ", rule.generate_traverse_shape(false, parse_nodes), "),
+                ) -> Self {" //-- start fn
+            );
+            if fields.is_empty() {
+                put!(out, name, "::", variant_name, "(Handle { node: _node, parser, _marker: PhantomData, })");
+            } else {
+                put!(out, name, "::", variant_name, " {"); // --start variant
+                for (field_name, paths) in fields {
+                    if rule.field_pathset_is_refutable(&paths) {
+                        put!(out, field_name, ": None.or(_r"); // -- start paren
+                        for path in paths {
+                            for p in path {
+                                put!(out, " .", p);
+                            }
+                        }
+                        put!(out, ").map(|node| Handle {
+                            node,
+                            parser,
+                            _marker: PhantomData,
+                        }),"); // -- end paren
+                    } else {
+                        put!(out, field_name, ": Handle { node: _r");
+                        assert_eq!(paths.len(), 1);
+                        for p in paths.get_index(0).unwrap() {
+                            put!(out, " .", p);
+                        }
+                        put!(out, ", parser, _marker: PhantomData, },");
+                    }
+                }
+                put!(out, "}"); // -- end variant
+            }
+            put!(out, "}"); // -- end fn
```

Comment:
Should we just move to quasi-quotation? Sadly @dtolnay is opposed to `${...}`-style nesting in quasi-quotation, which is the only way this code would look reasonable at all.